### PR TITLE
FusedAdam & CPUAdam

### PR DESCRIPTION
**Pull request by cozytk**
## Title

-FusedAdam & CPUAdam

## Description

- Implement FusedAdam & CPUAdam

## Tasks

- [x] Implement FusedAdam
- [x] implement CPUAdam
- [x] Test FusedAdam
- [x] Test CPUAdam
- [x] Test FusedSclaeMaskSoftmax (Name changed)

### CODE REVIEW

**Comment by hyunwoongko** [comment]
- Circular import occurs again and I temporarily commented out theUnion[ZeroRed...
  - remove this.
- DeepSpeedCPUAdam -> CPUAdam

**Comment by cozytk** [comment]
@hyunwoongko Done!

File: oslo/torch/_C/csrc/includes/gemm_test.h [hunk 1]
```diff
@@ -0,0 +1,214 @@
+
```
Comment (hyunwoongko) [inline oslo/torch/_C/csrc/includes/gemm_test.h:1]: Does this need?

File: oslo/torch/_C/csrc/includes/cublas_wrappers.h [hunk 2]
```diff
@@ -0,0 +1,36 @@
+#pragma once
```
Comment (hyunwoongko) [inline oslo/torch/_C/csrc/includes/cublas_wrappers.h:1]: Does this need?

File: oslo/torch/_C/csrc/includes/context.h [hunk 3]
```diff
@@ -0,0 +1,164 @@
+#pragma once
+
+#include "cublas_v2.h"
+#include "cuda.h"
+#include "curand.h"
+#include "gemm_test.h"
```
Comment (cozytk) [inline oslo/torch/_C/csrc/includes/context.h:6]: @hyunwoongko  `çpu_adam.h` requires `context.h`. `context.h` requires `gemm_test.h`

File: oslo/torch/_C/csrc/includes/gemm_test.h [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+#pragma once
+
+#include "StopWatch.h"
+#include "cublas_wrappers.h"
```
Comment (cozytk) [inline oslo/torch/_C/csrc/includes/gemm_test.h:5]: @hyunwoongko `gemm_test.h` includes `cublas_wrappers.h`

File: oslo/torch/_C/csrc/includes/context.h [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+  std::pair<uint64_t, uint64_t> IncrementOffset(uint64_t offset_inc) {
+    uint64_t offset = _curr_offset;
+    _curr_offset += offset_inc;
+    return std::pair<uint64_t, uint64_t>(_seed, offset);
+  }
+
+  void SetSeed(uint64_t new_seed) { _seed = new_seed; }
+
+  void TestGemmFP16(bool test_gemm, int batch_size, int seq_len, int head_num,
```
Comment (hyunwoongko) [inline oslo/torch/_C/csrc/includes/context.h:97]: You can remove this. no?

**Comment by cozytk** [comment]
- [x] remove type annotating `Union[ZeorRedu...]`
- [x] changed named to `CPUAdam`
- [x] checked `FusedAdam`, `CPUAdam` tests work

File: oslo/torch/_C/csrc/includes/context.h (hunk 5 above)
Comment (cozytk) [reply oslo/torch/_C/csrc/includes/context.h:97]: @hyunwoongko I'm gonna apply it!

**Comment by cozytk** [comment]
@hyunwoongko 
- [x] remove func `void TestGemmFP16`
- [x] remove headers `gems_test.h` `StopWatch.h` `cublas_wrappers.h`
- [x] add include `cuda_fp16.h` in `custom_cuda_layers.h`

**Pull request merged**